### PR TITLE
When TopologyManagerPolicy is None, skip checks in NewManager.

### LINKED ISSUE #116391: When TopologyManagerPolicy is None, there is no need to discover the NUMA topology.
Follow up from: https://github.com/kubernetes/kubernetes/pull/116093/files#r1128138681 See details there.

When `TopologyManagerPolicy` is `None`, there is no need to discover the NUMA topology.

/sig node
/priority backlog
/kind feature
/help-wanted

### PR DESCRIPTION
**Pull request by AxeZhan**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Calls to NewPolicyOptions and NewNUMAInfo calls should be skipped if the policy is PolicyNone
#### Which issue(s) this PR fixes:

Fixes #116391

#### Special notes for your reviewer:
It's my first time contribue to sig/node, I'm not very familiar with this part of the code. If something is wrong, please give me a hint and I'll dig into it.
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 1]
```diff
@@ -135,14 +135,21 @@ var _ Manager = &manager{}
...
+
+	if topologyPolicyName != PolicyNone {
+		opts, err = NewPolicyOptions(topologyPolicyOptions)
+		if err != nil {
+			return nil, err
+		}
+		numaInfo, err = NewNUMAInfo(topology, opts)
+		if err != nil {
+			return nil, fmt.Errorf("cannot discover NUMA topology: %w", err)
+		}
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:152]: this is the most straightforward approach and probably the best start.
But entering the `switch` below with `nil` values, which is unprecedented, gives me a fuzzy feeling.
We should explore if there are ways to rethink the flow in this function such like (brainstorming now) treating `PolicyNone` as special case and short circuiting out if it is.
If turns out there's no cleaner way to handle this case, let's go with the simple approach. But we should try first.

**Comment by ffromani** [comment]
unit lane failures are unrelated, they look scary and we will probably need to file a separate tracking issue:
```
 === Failed
=== FAIL: pkg/scheduler/framework/plugins/volumebinding TestVolumeBinding/zonal_volumes_with_close_capacity_are_preferred_(custom_shape) (0.08s)
W0319 09:15:03.112538   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112616   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112673   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112718   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112762   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112830   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112900   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
W0319 09:15:03.112982   60750 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage. 
```

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 1 above)
Comment (kidddddddddddddddddddddd) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:152]: we can do a quick check for policyName is none, and wrap a function to getScope by policy.
this way, we can remove the rest checks for `case PolicyNone` in `NewManager`, will this be better?
```
func NewManager() {
	if topologyPolicyName == PolicyNone {
		policy := NewNonePolicy()
		scope,err := getScope(topologyScopeName,policy)
		if err!=nil {
			return nil, err
		}
		manager := &manager{
			scope: scope,
		}
		return manager, nil	
	}
       //continue the logic
}

func getScope(topologyScopeName string, policy Policy) (scope Scope, err error) {
	switch topologyScopeName {
	case containerTopologyScope:
		scope = NewContainerScope(policy)

	case podTopologyScope:
		scope = NewPodScope(policy)

	default:
		return nil, fmt.Errorf("unknown scope: \"%s\"", topologyScopeName)
	}
	return scope, nil
}

```

**Comment by kidddddddddddddddddddddd** [comment]
> unit lane failures are unrelated, they look scary and we will probably need to file a separate tracking issue:

shoule be a flaky one, I can file a report later.

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 1 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:152]: I'll think about options and come back with suggestions. We totally should not wait for this fix too long, but let's also take some time to evaluate alternatives, the code is simple enough and well covered by tests so we have some room for exploration.

Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:152]: We could just add new scope called `scopeNone` which would just noop and return early like this:
```go
func NewManager(paramList) (Manager, error) {
    if topologyPolicyName == PolicyNone {
        return &manager{
            scope: scopeNone(),
        }, nil
    }
    ... // continue with other policies and scopes
}
```

Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:152]: Yes.

Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/topology_manager.go:152]: > We could just add new scope called `scopeNone` which would just noop and return early like this:
> 
> ```go
> func NewManager(paramList) (Manager, error) {
>     if topologyPolicyName == PolicyNone {
>         return &manager{
>             scope: scopeNone(),
>         }, nil
>     }
>     ... // continue with other policies and scopes
> }
> ```
> 
> Right now both `container` and `pod` scope has this check at the beginning of `Admit` func:
> 
> ```go
> func (s *containerScope) Admit(pod *v1.Pod) lifecycle.PodAdmitResult {
> 	// Exception - Policy : none
> 	if s.policy.Name() == PolicyNone {
> 		return s.admitPolicyNone(pod)
> 	}
> ```
> 
> we could get rid of those with this approach.

I tend to like this direction, I think it has potential. Let's evaluate the full change and see how it works.
Thanks for chiming in @PiotrProkop ! Unforetunately I was up until now more busy than expected prepping for kubecon eu.

File: pkg/kubelet/cm/topologymanager/scope.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ package topologymanager
 import (
 	"sync"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/scope.go:22]: remove this alias

File: pkg/kubelet/cm/topologymanager/scope_container.go [hunk 3]
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 package topologymanager
 
 import (
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/scope_container.go:20]: remove this alias

File: pkg/kubelet/cm/topologymanager/scope_none.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+type noneScope struct {
+	scope
+}
+
+// Ensure noneScope implements Scope interface
+var _ Scope = &noneScope{}
+
+// NewNoneScope returns a none scope.
+func NewNoneScope(policy Policy) Scope {
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/scope_none.go:33]: I think we don't have to pass the policy in here. It'll always be `none`

File: pkg/kubelet/cm/topologymanager/scope_none_test.go
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/scope_none_test.go:1]: are you planning to add tests? if not then it's better to remove this file.

File: pkg/kubelet/cm/topologymanager/scope_pod.go [hunk 5]
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 package topologymanager
 
 import (
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/scope_pod.go:20]: remove alias

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 6]
```diff
@@ -21,7 +21,7 @@ import (
 	"time"
 
 	cadvisorapi "github.com/google/cadvisor/info/v1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:24]: remove alias

File: pkg/kubelet/cm/topologymanager/topology_manager.go [hunk 7]
```diff
@@ -135,6 +135,11 @@ var _ Manager = &manager{}
 func NewManager(topology []cadvisorapi.Node, topologyPolicyName string, topologyScopeName string, topologyPolicyOptions map[string]string) (Manager, error) {
 	klog.InfoS("Creating topology manager with policy per scope", "topologyPolicyName", topologyPolicyName, "topologyScopeName", topologyScopeName)
 
+	if topologyPolicyName == PolicyNone {
+		policy := NewNonePolicy()
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/topology_manager.go:139]: remove this and just initiate it in `NewNoneScope`

File: pkg/kubelet/cm/topologymanager/scope_none_test.go
Comment (AxeZhan) [reply pkg/kubelet/cm/topologymanager/scope_none_test.go:1]: I'm considering. Do you think it's necessary to add tests for this struct. It seems it has nothing new and should be well covered in package `topologymanager` tests.

Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/scope_none_test.go:1]: I think we don't have to add unit tests for that as long as code coverage isn't decreased 😄

File: pkg/kubelet/cm/topologymanager/scope.go [hunk 8]
```diff
@@ -31,6 +31,8 @@ const (
 	containerTopologyScope = "container"
 	// podTopologyScope specifies the TopologyManagerScope per pod.
 	podTopologyScope = "pod"
+	// noneTopologyScope specifies the TopologyManagerScope when topologyPolicyName is none.
+	noneTopologyScope = "none"
```
Comment (ffromani) [inline pkg/kubelet/cm/topologymanager/scope.go:35]: will this be exposed or should this be kept as implementation detail? (I think the latter unless there are strong reasons to expose)

File: pkg/kubelet/cm/topologymanager/scope_none.go (hunk 4 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/scope_none.go]: Agreed, this is being treated as special case anyway

File: pkg/kubelet/cm/topologymanager/scope_none_test.go
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/scope_none_test.go:1]: With the proposed change hardcoding PolicyNone in scopeNone the code is so trivial it doesn't deserve any test

File: pkg/kubelet/cm/topologymanager/topology_manager.go (hunk 7 above)
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/topology_manager.go]: perhaps add a one-line comment emphasizing we that with policy=none the scope is not relevant, so we can short circuit here

File: pkg/kubelet/cm/topologymanager/scope_none.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+type noneScope struct {
+	scope
+}
+
+// Ensure noneScope implements Scope interface
+var _ Scope = &noneScope{}
+
+// NewNoneScope returns a none scope.
+func NewNoneScope() Scope {
+	return &containerScope{
```
Comment (PiotrProkop) [inline pkg/kubelet/cm/topologymanager/scope_none.go:34]: it should be `noneScope` instead of `containerScope`

Comment (AxeZhan) [reply pkg/kubelet/cm/topologymanager/scope_none.go]: Sorry, a copy mistake.

File: pkg/kubelet/cm/topologymanager/scope.go (hunk 8 above)
Comment (AxeZhan) [reply pkg/kubelet/cm/topologymanager/scope.go:35]: Currently I think there is no need to Expose this.

File: pkg/kubelet/cm/topologymanager/scope.go (hunk 2 above)
Comment (AxeZhan) [reply pkg/kubelet/cm/topologymanager/scope.go]: Seems some files in this package are using alias, and some others don't.Should we unify them?

**Comment by ffromani** [comment]
looks nice. I'll have the final review later. Could you please squash the fixing commit in the main one?

File: pkg/kubelet/cm/topologymanager/scope.go (hunk 2 above)
Comment (PiotrProkop) [reply pkg/kubelet/cm/topologymanager/scope.go]: my rule of thumb is to leave imports how it were, unless you are importing something new.

**Comment by AxeZhan** [comment]
> looks nice. I'll have the final review later. Could you please squash the fixing commit in the main one?

Some tests are failing, I'll look at them later. You can review after that .

**Review by ffromani — COMMENTED** [review]
/lgtm
/hold

(sorry for late review)
I like how the change turned out, thanks!
@PiotrProkop was also reviewing, please unhold when happy

**Review by klueska — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**